### PR TITLE
types: define NonNegativeInteger

### PR DESCRIPTION
**Pull request by kurtmilam**
I'm not able to run `make test lint` on my PC yet. The change is small and atomic, and I ran my tests for the new type using another method.

I'm unsure about the right way to treat `-0`. `$test( [], NonNegativeInteger, -0 )` returns `true`. I included a test for this in the test suite to make that clear.

The other candidate name for this Type is `NaturalNumber`. Unfortunately, there seems to be a good deal of controversy over whether zero is included in the set of natural numbers.

### CODE REVIEW

**Comment by kurtmilam** [comment]
A note on the usefulness: This Type is useful for declaring ranges and comparing or setting lengths.

File: index.js [hunk 1]
```diff
@@ -617,6 +617,15 @@
     function(x) { return RegExp_._test(x) && !x.global; }
   );
 
+  //# NonNegativeInteger :: Type
+  //.
+  //. Type comprising `0` and every [`PositiveInteger`][] value. Also known as
```
Comment (davidchambers) [inline index.js]: Let's be clear about `-0` (which I agree should be a member of the type):

```diff
-  //. Type comprising `0` and every [`PositiveInteger`][] value. Also known as
+  //. Type comprising `0`, `-0`, and every [`PositiveInteger`][] value. Also known as
```

File: index.js [hunk 2]
```diff
@@ -617,6 +617,15 @@
...
+  //. the set of Natural Numbers under ISO 80000-2:2009.
```
Comment (davidchambers) [inline index.js]: Is it necessary to capitalize _Natural Numbers_?

File: index.js [hunk 3]
```diff
@@ -617,6 +617,15 @@
...
+  var NonNegativeInteger = NullaryTypeWithUrl(
+    'sanctuary-def/NonNegativeInteger ',
```
Comment (davidchambers) [inline index.js]: `s/Integer '/Integer'/`

File: test/index.js [hunk 4]
```diff
@@ -1733,6 +1733,20 @@ describe('def', function() {
...
+    eq($.NonNegativeInteger.url, 'https://github.com/sanctuary-js/sanctuary-def/tree/v' + version + '#NonNegativeInteger');
+
+    function isNonNegativeInteger(x) {
+      return $.test($.env, $.NonNegativeInteger, x);
+    }
+    eq(isNonNegativeInteger(0), true);
+    eq(isNonNegativeInteger(-0), true);
+    eq(isNonNegativeInteger(3.14), false);
+    eq(isNonNegativeInteger(1), true);
+    eq(isNonNegativeInteger(new Number(1)), false);
```
Comment (davidchambers) [inline test/index.js]: Let's also test a negative integer.

**Comment by davidchambers** [comment]
`NonNegativeInteger` needs to be exported. ;)

File: index.js [hunk 5]
```diff
@@ -617,6 +617,15 @@
     function(x) { return RegExp_._test(x) && !x.global; }
   );
 
+  //# NonNegativeInteger :: Type
+  //.
+  //. Type comprising `0`, `-0`, and every [`PositiveInteger`][] value.
```
Comment (davidchambers) [inline index.js]: According to [`make lint`][1] there's a missing link definition. I suspect there's no link definition for ``[`PositiveInteger`]`` at the bottom of this file. Please add one. :)


[1]: https://circleci.com/gh/sanctuary-js/sanctuary-def/792

File: index.js [hunk 6]
```diff
@@ -617,6 +617,15 @@
...
+  //. Also known as the set of natural numbers under ISO 80000-2:2009.
+  var NonNegativeInteger = NullaryTypeWithUrl(
+    'sanctuary-def/NonNegativeInteger',
+    function(x) { return PositiveInteger._test(x) || x === 0; }
```
Comment (davidchambers) [inline index.js]: It just occurred to me that `Integer._test(x) && x >= 0` may be a clearer expression. What do you think?

We could also update the description:

_Type comprising every non-negative [`Integer`][1] value (including `-0`)._


[1]: https://github.com/sanctuary-js/sanctuary-def#Integer

Comment (kurtmilam) [inline index.js]: I'm good with that. I had actually changed the test to
```JavaScript
function(x) { return Integer._test(x) && x >= 0; }
```
earlier, then waffled, thought about asking you and finally changed it back.

I guess it makes more sense to choose the inclusive type and add a filter than it does to choose the exclusive type and add an exception to the rule.

File: index.js [hunk 7]
```diff
@@ -617,6 +617,15 @@
     function(x) { return RegExp_._test(x) && !x.global; }
   );
 
+  //# NonNegativeInteger :: Type
+  //.
+  //. Type comprising every non-negative Integer value (including -0).
```
Comment (davidchambers) [inline index.js]: Let's wrap `Integer` and `-0` in backticks, and have the former link to `#Integer`.

File: index.js [hunk 8]
```diff
@@ -2632,6 +2642,7 @@
 //. [`Number`]:             #Number
 //. [`Object.create`]:      https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create
 //. [`Pair`]:               #Pair
+//. [`PositiveInteger`]:    #PositiveInteger
```
Comment (davidchambers) [inline index.js]: This is no longer necessary.

Comment (kurtmilam) [inline index.js]: I had removed it, then it came back in the merge. I've already removed it again (although haven't committed/pushed). I'm something of a git newb when it comes to using it collaboratively

File: index.js (hunk 7 above)
Comment (kurtmilam) [inline index.js]: Will do. I'll get used to the house style and requirements soon, hopefully :)

File: index.js (hunk 8 above)
Comment (kurtmilam) [inline index.js]: I'm trying to cut down on the noise via `commit --amend` and `rebase`, but I fear I may be making things worse 😆

**Comment by davidchambers** [comment]
The changes look good.

I'd like to squash the commits before merging. I rebased the commits to save you the trouble: <https://github.com/sanctuary-js/sanctuary-def/commit/4fb24885fcbe5ca47080edf3176a67886accb8e9>. You should be able to update the pull request by running these commands:

```console
$ git checkout master
$ git branch -m add-non-negative-integer-type{,.orig}  # save your existing branch
$ git checkout -b add-non-negative-integer-type
$ git pull git@github.com:sanctuary-js/sanctuary-def.git kurtmilam/add-non-negative-integer-type
$ git commit --amend  # make yourself the committer
$ git push origin add-non-negative-integer-type --force
```

In future I suggest avoiding merge commits on feature branches when contributing to Sanctuary projects, as they make squashing more time-consuming.

**Comment by kurtmilam** [comment]
Thanks for the assistance. In order to avoid merge commits when amending or rebasing, is it OK to force push to the branch in my fork? I think that's where the merges came from. I tried to squash my commits but didn't force push.

**Comment by davidchambers** [comment]
> In order to avoid merge commits when amending or rebasing, is it OK to force push to the branch in my fork?

Absolutely. You're free to do as you please on your fork. For shared repositories I only `--force` push to "my" branches (the ones whose names begin `davidchambers/`).
